### PR TITLE
Operator

### PR DESCRIPTION
**Pull request by badochov**
Implement golang version of a subset of the `operator` library

### CODE REVIEW

File: src/operator/operator.go [hunk 1]
```diff
@@ -0,0 +1,746 @@
...
+
+import (
+	"fmt"
+	"github.com/gojinja/gojinja/src/utils/slices"
+	"math"
+	"reflect"
+	"strings"
+)
+
+func ToInt(v any) (i int64, ok bool) {
```
Comment (jmolinski) [inline src/operator/operator.go]: I'm confused; isn't it a copy&paste from #19 ?

Comment (jmolinski) [reply src/operator/operator.go]: Other methods too

File: src/operator/operator.go [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+func ToComplex(v any) (c complex128, ok bool) {
+	if p, ok := v.(complex64); ok {
+		return complex128(p), ok
+	}
+	c, ok = v.(complex128)
+	return
+}
+
+func IsNumeric(v any) bool {
+	_, isInt := ToInt(v)
```
Comment (jmolinski) [inline src/operator/operator.go]: maybe

```go
	_, isInt := ToInt(v)
	if !isInt {
	   _, isFloat := ToFloat(v)
	   if !isFloat {
	        _, isComplex := ToComplex(v)
	       return isComplex
	    }
	}
	return true
}
```

to shortcircuit? most argumemts passed here will most likely be ints, so it'd skip checking for float/complex

File: src/operator/operator.go [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+func Add(a any, b any) (any, error) {
+	if iadd, ok := a.(IAdd); ok {
+		return iadd.Add(b)
+	}
+	if irAdd, ok := b.(IRAdd); ok {
+		return irAdd.RAdd(b)
+	}
+	if IsNumeric(a) && IsNumeric(b) {
+		return addNumeric(a, b), nil
+
```
Comment (jmolinski) [inline src/operator/operator.go]: not needed empty line

File: src/operator/operator.go [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+		return i.Sub(b)
+	}
+	if ir, ok := b.(IRSub); ok {
+		return ir.RSub(b)
+	}
+	if IsNumeric(a) && IsNumeric(b) {
+		return subNumeric(a, b), nil
+	}
+
+	return nil, fmt.Errorf("given elements are not subable")
```
Comment (jmolinski) [inline src/operator/operator.go]: subable? 😄

File: src/operator/operator.go [hunk 5]
```diff
@@ -0,0 +1,746 @@
...
+	}
+	if bothString(a, b) {
+		return a.(string) + b.(string), nil
+	}
+	if slices.Contains([]reflect.Kind{reflect.Slice, reflect.Array}, reflect.TypeOf(a).Kind()) &&
+		slices.Contains([]reflect.Kind{reflect.Slice, reflect.Array}, reflect.TypeOf(b).Kind()) {
+		return addSlices(a, b), nil
+	}
+
+	return nil, fmt.Errorf("given elements are not additible")
```
Comment (jmolinski) [inline src/operator/operator.go]: additive?

File: src/operator/operator.go [hunk 6]
```diff
@@ -0,0 +1,746 @@
...
+		return i.Div(b)
+	}
+	if ir, ok := b.(IRDiv); ok {
+		return ir.RDiv(b)
+	}
+	if IsNumeric(a) && IsNumeric(b) {
+		return divNumeric(a, b)
+	}
+
+	return nil, fmt.Errorf("given elements are not divable")
```
Comment (jmolinski) [inline src/operator/operator.go]: divisible

File: src/operator/operator.go [hunk 7]
```diff
@@ -0,0 +1,746 @@
...
+		return i.Pow(b)
+	}
+	if ir, ok := b.(IRPow); ok {
+		return ir.RPow(b)
+	}
+	if IsNumeric(a) && IsNumeric(b) {
+		return powNumeric(a, b), nil
+	}
+
+	return nil, fmt.Errorf("given elements are not powable")
```
Comment (jmolinski) [inline src/operator/operator.go]: honesty, who wrote these comments? xD

File: src/operator/operator.go [hunk 8]
```diff
@@ -0,0 +1,746 @@
...
+			if err != nil {
+				return false, err
+			}
+			if r == true {
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+	return false, fmt.Errorf("elements are not cointainable ")
```
Comment (jmolinski) [inline src/operator/operator.go]: elements don't support `in` operator

File: src/operator/operator.go [hunk 9]
```diff
@@ -0,0 +1,746 @@
...
+		reflect.Copy(c.Slice(i*v.Len(), (i+1)*v.Len()), v)
+	}
+	return c.Interface()
+}
+
+func multiplyNumeric(a any, b any) any {
+	res, _ := opNumeric(a, b, func(a any, b any) (any, error) {
+		switch v := a.(type) {
+		case int64:
+			return v * b.(int64), nil
```
Comment (jmolinski) [inline src/operator/operator.go:445]: what if b is in fact uint64? won't the type assert fail here?

**Review by jmolinski — CHANGES_REQUESTED** [review]
Good work, very important for the project

Please fix error comments - looks like they've been automatically generated?

Please add just a few simple test cases so that we're confident it works as expected

File: src/operator/operator.go (hunk 9 above)
Comment (badochov) [reply src/operator/operator.go:445]: THe way opNumeric is structured guarantees that a b are of the same type

File: src/operator/operator.go (hunk 7 above)
Comment (badochov) [reply src/operator/operator.go]: me myself and i

Comment (badochov) [reply src/operator/operator.go]: nto sure about the mental process but happened but i figured let's do it like name of function + able everywhere xD

File: src/operator/operator.go (hunk 1 above)
Comment (badochov) [reply src/operator/operator.go]: It is

**Comment by badochov** [comment]
Written test for Le, Add, Mul, Eq created #26 to write remaining tests

File: src/operator/operator.go (hunk 1 above)
Comment (jmolinski) [reply src/operator/operator.go]: Is that intentional? Will be deduplicated in the future?

**Review by jmolinski — APPROVED** [review]

File: src/operator/operator.go (hunk 1 above)
Comment (badochov) [reply src/operator/operator.go]: I unified the code in newest Commit

**Review by jmolinski — APPROVED** [review]

**Pull request merged**